### ISSUE TITLE
Provide a documented way to customize slugs generated by `SlugInput` / urlify

### ISSUE THREAD

**Issue by drimacus182** [wagtail/wagtail#11916]
I've just tried new 6.1 and as I can see `window.URLify` is deprecated and is not used by `SlugInput` internally. 

The problem is that current SlugInput urlify function is although universal but not entirely correct for every language. For instance, for Ukrainian it produces incorrect results according to transliteration rules:

<img width="545" alt="sluginput_ukrainian" src="https://github.com/wagtail/wagtail/assets/3090883/0683adfd-cd5c-43b1-9408-4e0540c05634">

Prior to 6.1 I was solving this using a tailored `URLify` function injected into admin. Now it lives inside a stimulus controller on a client js side, which is not straightforward to customize.

It would be great to have a documentation describing best practices on how this could be achieved.

**Comment by lb-** [comment]
Thanks for the report @drimacus182  - while the previous approach to override the global function `window.URLify` was never documented I can see how having some capability to override the behaviour here would be helpful.

I think we should try to consider a few ways to resolve this.

## 1. Potentially incorrect transliteration

You will have to help us out here, could you please advise the exact characters , in text (not just a screenshot), that you think should be updated to be transliterated differently?

All of this transliteration is based on the Django [`urlify.js`](https://github.com/django/django/blob/main/django/contrib/admin/static/admin/js/urlify.js) code.

We have now pulled this into a more manageable config file [`urlify.config.json`](https://github.com/wagtail/wagtail/blob/main/client/src/utils/urlify.config.json) as part of our refactor.

Maybe we should just try to fix this in our code (plus we could suggest a similar fix in Django's code), so that the characters are handled better.

## 2. Another unofficial work around

It should be possible to add an event listener to the slug input against the event `'w-sync:apply'` and then stop propagation and re-dispatch that same event with the desired transliteration.

This is not as clean as just mutating the global but should be enough to get you started with specific transliteration requirements.

This would still be an unofficial work around though, and could break in any future release.

```js
# .../static/js/admin.js

document.addEventListener(
  'w-sync:apply',
  (event) => {
    // return if the event is not from the slug field
    if (event.target.name !== 'slug') return;
    // return if the event has already been processed
    if (event.detail.processed) return;

    console.log('w-sync:apply', event);

    // stop the current event from propagating
    event.preventDefault();
    event.stopPropagation();

    // get the value to apply and mutate to suit our needs, dispatch a new event
    const value = event.detail.value || ''; // will be the original title's value
    event.target.dispatchEvent(
      new CustomEvent('w-sync:apply', {
        detail: {
          // change any values from the title field before it gets to the slugify controller
          value: value.replaceAll('ґ', 'g'),
          processed: true,
        },
      }),
      { bubbles: true, cancelable: true },
    );
  },
  { capture: true }, // ensure this listener captures the event first
);
```

```py
# ... /wagtail_hooks.py
from django.utils.safestring import mark_safe
from django.templatetags.static import static

from wagtail import hooks

@hooks.register('insert_global_admin_js')
def global_admin_js():
    src = static('js/admin.js')
    return mark_safe(f'<script src="{src}"></script>')
```

## 3. Add ability to intercept the urlify/slugify more easily with events

We should probably dispatch a change event when the slugify/urlify triggers, this could be intercepted easily and any further mutations of the field's value can be made.

Additionally we could add custom events for when urlify/slugify are **about** to happen and allow this value to be changed similar to how we do for the image/document title field. https://docs.wagtail.org/en/latest/advanced_topics/images/title_generation_on_upload.html#images-title-generation-on-upload

----

I would rather us try to get the transliteration right first though, but I am not sure what wider input is needed here from the community.

Generally transliteration is a error-prone process. We currently run into a similar problem with this for the form builder usage of [`cautious_slugify`](https://github.com/wagtail/wagtail/blob/main/wagtail/coreutils.py#L115), where some users want this to work like transliteration but some cases are not handled.

**Comment by lb-** [comment]
I have also posted in the `#translators` chat to get some other thoughts on all this.

For now @drimacus182  - can you test the work around above and see if that gets you to a point where you can use 6.1 in your production code. Any other suggestions welcome.

**Comment by drimacus182** [comment]
Hi @lb-, thanks for looking into this. I'll try to add my thoughts:

## Approach 1

I understand the desire to make slugs work out of the box, but I see a problem with this universal approach. The existing Django's [urlify.js](https://github.com/django/django/blob/main/django/contrib/admin/static/admin/js/urlify.js) and wagtail's [urlify.config.json](https://github.com/wagtail/wagtail/blob/main/client/src/utils/urlify.config.json) you've mentioned are just mapping letters one by one, without any knowledge of what language the original string is in. E.g. for a Cyrillic character, it starts with `RUSSIAN_MAP` and proceeds to `UKRAINIAN_MAP` only if the character is not from the Russian alphabet. As a result, we get correct transliteration for Russian, but partially incorrect for Ukrainian. I mean, there are letters that are common to Russian and Ukrainian but have different phonetics in those languages and hence different transliterations. I suppose the same is true for other Cyrillic languages like Belarusian, Bulgarian, Serbian etc.

Example from a `RUSSIAN_MAP`:
```js
["г", "g"], // should be "h" in Ukrainian
["и", "i"], // should be "y" in Ukrainian
```


Apart from phonetics, there are different transliteration standards possible for different contexts and languages. So there is no simple "one-size-fits-all" solution for every use case. Such a solution would require knowledge of what language the original string is in and will introduce lots of complexity. The official transliteration standard for Ukrainian [can be found here](https://czo.gov.ua/en/translit) 

## Approach 2
I'm generally fine with the 2nd solution, and thank you for the complete code sample. I'd try to use this with my deployment, but can't promise to do this quickly. It feels a bit hacky for me though, especially considering it is unofficial and can break in future releases. 


## Approach 3
>we could add custom events for when urlify/slugify are about to happen and allow this value to be changed 

It sounds most satisfying to me since my initial need was to have the ability to modify the internals of `SlugController.urlify` method in a documented and reliable way. As I understand, it won't be the "modification" of the internal slugify method, but rather an expected placeholder to inject a custom slugify logic. I'm new to Stimulus, and can't fully understand everything that's happening inside controllers, but I suppose the value from the title input is passed to the slug input as is, and the event handler you propose will be able to get an original unmodified string and set the slugified value.


To summarise, I think it's straightforward for a developer to write a transliteration function that will fit all their needs. On the contrary, it's almost impossible to create a simple transliteration logic that will make everyone happy. So the solution should be any of those that allow a developer to define their custom logic.

**Comment by lb-** [comment]
@drimacus182 here's a proposed client-side API for overriding this behaviour. 


```js
document.addEventListener('w-clean:urlify', event => {
  const {
    apply /* A function you can use, once, to apply a custom value */,
    currentValue /* Current value (e.g. 'рецепт чізкейку') */,
    newValue /* Default transliteration (e.g. 'hecept czizkejku' ) */
  } = event.detail;

  // If you want to stop the new value from being used and make your own...
  event.preventDefault();
  apply(myCustomFunction(currentValue) /* e.g. 'recept czizkejku' */);
});
```

A similar event would be available for `w-clean:slugify`. These give you two levels of control, first - just stopping the change all together, second - applying a custom value.

Additionally, we would add change events to the slug field, so you can listen for a change **after** it's been applied and do something else then.

This approach leverages the DOM events API, avoids globals and gives us room to enhance later with additional data being provided to the event. You'll also have access to the input with `event.target` if you want to do more complex things.

However, this would not easily allow any Async usage, as the preventDefault call needs to be used. It would be possible to still use this method to do a fetch call and update the field after the response comes back though, just may be a bit more involved. 

Finally, this involves us updating the `SlugController`/`w-slug` to `CleanController`/`w-clean` which I think is a better name for it and it lets us add new behaviour to this controller later.

**Comment by ismayil-ismayilov** [comment]
If possible, please also consider Azerbaijani letters. This would also cover many Turkic based languages at once (including Turkish). Azerbaijani has more letters than others in average.
Here is the map:
```
"AZERBAIJANI": [
  ["ə", "e"],
  ["Ə", "E"]
  ["ı", "i"],
  ["I", "I"],
  ["ö", "o"],
  ["Ö", "O"],
  ["ü", "u"],
  ["Ü", "U"],
  ["ğ", "g"],
  ["Ğ", "G"],
  ["ç", "c"],
  ["Ç", "C"],
  ["ş", "s"],
  ["Ş", "S"],
]
```

**Comment by laymonage** [comment]
Perhaps instead of [`urlify.config.json`](https://github.com/wagtail/wagtail/blob/main/client/src/utils/urlify.config.json), we should inline it as a Python dictionary and allow developers to extend it via a hook/setting? From there, we use `json_script` to render it in the template, to then be parsed by the URLify function.

**Comment by lb-** [comment]
I like that idea @laymonage - we could still allow the JS override in the future but I think having a non-JS way to do simpler customisations would be a better approach initially. 

Could you flag this for the core team discussion, see if there are any other ideas/inputs.

**Comment by allcaps** [comment]
I've made a suggestion on Slack, and cross-post it here to keep the ref:

> Would it be an idea to have Wagtail provide a slugify endpoint? This way, the frontend could be 'dumb', and the backend could have a new Wagtail hook to override the slugify behaviour.
This allows a "global" slugify function that both frontend and backend can use. Keeping slugify results the same between front and back.
It would also be straightforward for integrators to customise, and easy for Wagtail to support.

See Slack for the full discussion: https://wagtailcms.slack.com/archives/CTKN6UXKN/p1714721188163579?thread_ts=1714686970.068729&cid=CTKN6UXKN

**Comment by Stormheg** [comment]
The use case seems common enough to me to warrant a documented approach indeed. If we go the custom code route, I would prefer a custom urlify config file over custom JS in the admin for simplicity.

Not convinced that a universal approach is not feasible. If the mapping of characters to the incorrect language raised by [drimacus182](https://github.com/drimacus182) is the only issue, forgive my naïveté, but don't we already have information about the original locale available? Example: if the editor of a site configured for Ukrainian created a page, the locale of that page would be Ukrainian. We can use this information to consult the Ukrainian character map first. If it's a Russian site we would use the Russian character map first. Wouldn't that solve the issue? No need for customizing, it would work out of the box. Or are there other issues I'm not aware of? Would love to hear your input @drimacus182.

I don't think making the slugify feature a little smarter would introduce _that_ much extra complexity to support. We'd restructure the urlify config to support looking up maps by language code, pull the current language code from the page and use that as 'preferred locale' when transliterating. All easy things to say from the comfort of my chair, I'm not familiar with `SlugInput` or its inner workings.

---

We could decide we don't want the added complexity and leave it up to developers to implement their own slugify mechanism, but that seems a little unfair to me. I believe it should work correctly out of the box if at all feasible.

**Comment by laymonage** [comment]
@Stormheg I agree that changing the mapping to use the locale's language code as the key is a better approach. It may result in a bigger character map especially if there are multiple languages that could utilise the same mapping, though, so if we care about the dictionary size then perhaps we could also add support for aliases to let multiple languages use the same mapping.

However, I think another issue is that Wagtail's character map may be incomplete. It's true that anyone can submit a pull request to add support for more characters/languages, but that means they'd have to wait until the next release to be able to use it. If there was a way to customise the mapping, developers can do it in the mean time until they upgrade for the next release.

**Comment by Stormheg** [comment]
We discussed this during the core team meeting 

Summarizing;

- Using the wrong character map to transliterate is a bug. We should fix that. Preferably using a library that already has this solved and contains additional transliteration maps that may currently be missing from our `urlify.config.json`.
- We are a little divided about whether or not to officially support customising the slugify feature. I'm in the nay-camp, my personal take on it is that it wouldn't be used much because I believe people would only customize it when they encounter a bug that should be fixed in Wagtail. People can resort to unoffical JS workarounds to bridge the gap until a patched release.

**Comment by gasman** [comment]
I agree with @Stormheg. Delivering a solution that "just works" for Ukrainian transliteration solves the immediate problem - if any other situations arise in future where a custom function is necessary, we can cross that bridge when we come to it. In the meantime, the workaround that @lb- has provided should ensure that this is not a blocker for anyone upgrading to 6.1, so I'll bump the milestone to 6.2.

**Comment by drimacus182** [comment]
@Stormheg 
>We'd restructure the urlify config to support looking up maps by language code, pull the current language code from the page and use that as 'preferred locale' when transliterating

I'm generally fine with this solution and will use it in my deployment if Wagtail team is comfortable creating and maintaining character maps for different languages.  

What I'm a bit concerned with is the lack of fine control for a developer to make customisations. I can imagine scenarios when you may want slightly different transliteration logic for Pages and Snippets (i.e. blog posts and authors). And some other urlify customisations like removing stop words. 

These are a bit hypothetical though, and can be addressed using the workaround @lb- has provided above.

**Comment by drimacus182** [comment]
Hi @lb-, I've finally tried the workaround using `'w-sync:apply'` you provided. And I'm getting the same behaviour as described in this issue https://github.com/wagtail/wagtail/issues/10532. It stops updating the slug after tab change.

I’d be grateful if you could take a look at this

**Comment by drimacus182** [comment]
I understand this issue has been moved to a future release, and I appreciate the workaround provided by @lb- above (2nd point from https://github.com/wagtail/wagtail/issues/11916#issuecomment-2091807670). Unfortunately, the workaround doesn’t seem to work for me, at least for Cyrillic titles.

The issue appears to be identical to the one described here (https://github.com/wagtail/wagtail/issues/10532), which has already been fixed. It seems related to the stimulus controllers, but my understanding of them is very limited, and I’m unable to investigate further.

As a result, I’m currently unable to upgrade beyond Wagtail 6.0. I would greatly appreciate any help with resolving this issue or fixing the workaround, as it would allow me to migrate to the latest versions of Wagtail

**Comment by lb-** [comment]
Hey all. I'll try to have another look at this and see what we can do.

**Comment by lb-** [comment]
**Quick update**

OK, the reason the JS workaround only works the 'first' time is due to the behaviour where on a new page load the JS will check the expected slugified value against the title value. If the value is different to what's expected, then the sync as you type will no longer happen.

This is intentional as we assume if the user has manually edited the slug to be different to the converted title they have manually edited it and we should not try to be 'smart' for the user in this case.

I had not considered that nuance when suggesting the original work around.

**Comment by lb-** [comment]
OK we will not simply be able to 'fix' the Ukrainian characters it seems, the way the Django transliteration port works is that it deeply merges character mapping.

If we add the different Ukraine characters that are different to Russian, this will mean that they will override the Russian ones as Ukraine characters are after them in the array.

Will add another update when I have a bit of a new plan to suggest.

**Comment by lb-** [comment]
I have a PR up with an incremental improvement here, this is a complex problem though.

The PR will basically add a way to override urlify/slugify values with DOM event listeners, moving to support async resolving of these 'cleaned' values.

I have not included any changes to the `UKRAINIAN_MAP` in `client/src/utils/urlify.config.json` as this will just break other languages and I feel it's not the right way to solve this problem.

I think we should provide a documented mechanism to override the urlify behaviour unless we (Wagtail) are willing to fully own the complexity of this.

## Ideas (better customisation)

1. Client-side - PR created that provides a JS basic customisation option.
2. Client-side - We potentially provide the base URLify function and/or the downcode mapping to be also overridden.
3. Server-side - Supply the downcode mapping from the server (as a once off, per page, request similar to icon sprites), could be overridden with hooks, or even be be dynamic per page locale (harder, means it cannot be cached).
4. Server-side - Make an admin API call every single change, handle all conversion server-side, including hooks. This is likely going to impact performance and we would need to consider multiple in flight requests, cancellation and error handling.

## Ideas (enhance existing urlify conversion)

This is a bit of a minefield of complexity, broadly we would need to consider the intended locale (source language), what kind of conversion (transliteration/transcription), nuances of the position of characters and more. On top of that we would be moving completely away from our current 'copy Django' approach.

We could consider using something like [`anyascii`](https://github.com/anyascii/anyascii) on the client or server. Not currently used for user facing strings though.

Any major change to the current conversion would likely end up being a per language complex Regex solution along with a breaking change as many existing slugs would be 'out of sync' with the raw title after the update and no longer automatically change as the title changes.

----

I'll hopefully get the chance to discuss this at the core team meeting today.

**Comment by lb-** [comment]
General feedback at core team was that this interim approach seems sound, we probably need to consider more broadly how much robust support for urlify should be built in.

There was consensus that we should look at future enhancements that **take the active locale into consideration**, but this may not be done quickly or easily.

Further to that, the desire is to keep logic for this on the client, so that it's not hitting the admin server for every slug change. There may still be a desire to have the initial mapping come from the server though.

For now, we will try to get the PR in (still needs core team reviews & some questions answered).

@drimacus182 - I would appreciate your look at https://github.com/wagtail/wagtail/pull/12379 for a potential interim solution. Note that if you do need the locale for the page that's being edited it will be available at `window.wagtailConfig.ACTIVE_CONTENT_LOCALE`.

Do you feel this PR, if implemented, would give you enough tools to override things nicely, or would you still want access to the 'original' urlify function or downcode mapping?

**Comment by drimacus182** [comment]
Hi @lb-. Thank you so much for looking into this so quickly.

Regarding your question, I’d be fine with any solution that allows me to implement a custom urlify behaviour. As long as I have the flexibility to override it as needed, that would work for me.

**Comment by lb-** [comment]
Pulling over comment from @gasman on #12379

> Discussed at the core team meeting today.
> 
> My gut feeling is that an event listener is the wrong operating model for this. An event is a one-to-many broadcast, indicating "this action is happening, and we're giving external components the opportunity to respond to it (including potentially modifying it)" - but in this case we're not so much responding to the action, as assigning a single component to be responsible for _performing_ the action.
> 
> As with previous discussions logged at #11916, there's still some scepticism within the team over whether a general-purpose function override is the best way to handle this - if the one established use case for this is handling languages such as Ukrainian (where the 'global' character mapping isn't usable), then being able to supply a custom mapping table would do the job. (In the longer term, we would hope to incorporate these language-specific mappings into Wagtail itself, enabling the appropriate ones in response to the active locale, so that it wouldn't be necessary for individual sites to maintain their own customisations.)

Thanks Matt, happy with this generally.

One caveat is that there will not ever be a language mapping that's suitable, a mapping (character replacement) may be good enough but will never cover all cases (as commented above). Once you get into the low level detail there are nuances such as 'start of word, start of sentence, preceding characters' that all could change the transliterated character.

But, happy to give another go at this with a different tact if I have time, I will close the PR now and raise a new PR with just the 'clean up parts' for that controller.

**Comment by lb-** [comment]
@gasman  - OK, here's another proposal (with a [POC branch available](https://github.com/lb-/wagtail/tree/poc/11916-enhanced-slug-input-customisation)).

Would love your input also @allcaps @laymonage @drimacus182 @Stormheg 

This approach means that our base behaviour would consider the Page's target locale when processing slug/url transliteration values, covering common cases for cross-language differences in characters.

We would also allow a more complex use case, providing a RegEx replace mechanism that would run BEFORE the urlify code runs, allowing more nuanced replacements such as specific words, start of line or whatever we want. We could NOT document this initially, but the important thing is that this would be at the Django widget level, not the JavaScript level.

We would not have to reach for a way more fragile ajax approach and could provide a lot of customisability in a declarative way on the widget. Providing two layers of customisation (a. override the locale on the widget, if the active content locale is not suitable, b. allow for pre-cleaning of the value before it's urlified by regex/replace pairs).

---


## 1. Provide two-ish new kwargs to `SlugInput`

1. `locale` - expects an instance of `Locale`, with a `language_code` available
2. `replace` - a List of tuples/pairs that allow for arbitrary RegEx replacement sets
3. `flags` - maybe we allow this, provide the base RegEx flags, or we could infer this by finding the flags from the provided RegEx instances, throwing a warning if they are different. More exploration needed probably into how Python's `.pattern` resolves to a string when flags are provided & how JS parses these strings.

**We would patch in the active content (for pages) locale where appropriate.**


## 2. Enhance `urlify` to be smarter about how the mappings get resolved

Currently the [mapping order is hard-coded](https://github.com/wagtail/wagtail/blob/main/client/src/utils/urlify.config.json) based on the original Django port of these items, this means it's **impossible** right now to have any nuance across language differences.

By accepting the locale (or language code, not sure on best variable terminology), we can re-order the mapping to ensure the matched language entries so that these come in 'last'.

We would need to move further away from the Django port and also ensure we have a way to add the language codes to mapping entries. Something like `"UKRAINIAN_MAP":` would become `"UKRAINIAN_MAP:uk":` where the `:` separator is used to allow for multiple language codes, we would not need these for all mappings, just the language specific ones.

In addition, we would further refine the mapping to account for characters that are duplicated in the Russian/Ukrainian mappings but have different target values. As noted above, doing this now will cause no actual change due to the ordering of the characters.

The nice thing about this approach is it should allow for a backwards compatible change (if we want to go this direction).

## 3. Enhance `CleanController` to support the above

Now we enhance the `CleanController` to read out the values, do the initial replacing before urlify/slugify is called.

1. Add `replace`  & `replaceFlags` values
2. Add a new method `replace`
3. Work on performance a bit, we only want to be preparing the RegExp instances / validating these once
4. Add events dispatching (but not cancelling, this allows for better cross-controller coordination)


## 4. Documentation & upgrade considerations

If developers want to **opt out** of the locale behaviour, they could pass in a dummy `Locale` instance that returns a language code of 'false' (or similar).

Beyond that, they should not need to change any usage of `SlugInput` already in place (I think).

If we want to officially document  the replace mechanism on SlugInput or the underlying CleanController usage, we would also be in a position to solve https://github.com/wagtail/wagtail/issues/7417

**Comment by lb-** [comment]
We discussed the approach above at Core Team today and the general consensus was that this is a good approach.

The regex approach may be a bit 'too' much power in the Django side and maybe we want to make it easier to override/extend the controller class instead.

We agreed that the `SlugInput` kwargs approach with the regex is probably a good way to go, but maybe review the controller structure for easy 'method isolation' anyway.

**Comment by lb-** [comment]
@drimacus182  could you please cross check this unit test for me, hopefully this is correct. I have done my best to cross check against existing transliteration and the link you provided https://czo.gov.ua/en/translit


```js
it('should allow a supplied locale to re-order downcode mapping', () => {
  // mix of characters that are transliterated differently across Russian/Ukrainian
  const chars = 'Ц-йи-Ґ-Х-щ';
  const ruTransliteration = 'c-ji-g-h-sh';
  const ukTransliteration = 'ts-iy-g-kh-shch';
  expect(urlify(chars)).toEqual(ruTransliteration); // by default the Russian values come through with priority
  expect(urlify('Ц-йи-Ґ-Х-щ', { locale: 'ru' })).toEqual(ruTransliteration);
  expect(urlify('Ц-йи-Ґ-Х-щ', { locale: 'uk' })).toEqual(ukTransliteration);

  const title = 'жйи Проєкт Герой Їжак Георгій';
  const ruTransliteration2 = 'zhji-proyekt-geroj-yizhak-georgij';
  const ukTransliteration2 = 'zhiy-proyekt-heroi-yizhak-heorhii';
  expect(urlify(title)).toEqual(ruTransliteration2); // default ordering
  expect(urlify(title, { locale: 'ru' })).toEqual(ruTransliteration2);
  expect(urlify(title, { locale: 'uk' })).toEqual(ukTransliteration2);
});
```

And here's the updated configuration, only the Ukranian character mapping has changed.

```json
  {
    "RUSSIAN_MAP:ru": [
      ["а", "a"],
      ["б", "b"],
      ["в", "v"],
      ["г", "g"],
      ["д", "d"],
      ["е", "e"],
      ["ё", "yo"],
      ["ж", "zh"],
      ["з", "z"],
      ["и", "i"],
      ["й", "j"],
      ["к", "k"],
      ["л", "l"],
      ["м", "m"],
      ["н", "n"],
      ["о", "o"],
      ["п", "p"],
      ["р", "r"],
      ["с", "s"],
      ["т", "t"],
      ["у", "u"],
      ["ф", "f"],
      ["х", "h"],
      ["ц", "c"],
      ["ч", "ch"],
      ["ш", "sh"],
      ["щ", "sh"],
      ["ъ", ""],
      ["ы", "y"],
      ["ь", ""],
      ["э", "e"],
      ["ю", "yu"],
      ["я", "ya"],
      ["А", "A"],
      ["Б", "B"],
      ["В", "V"],
      ["Г", "G"],
      ["Д", "D"],
      ["Е", "E"],
      ["Ё", "Yo"],
      ["Ж", "Zh"],
      ["З", "Z"],
      ["И", "I"],
      ["Й", "J"],
      ["К", "K"],
      ["Л", "L"],
      ["М", "M"],
      ["Н", "N"],
      ["О", "O"],
      ["П", "P"],
      ["Р", "R"],
      ["С", "S"],
      ["Т", "T"],
      ["У", "U"],
      ["Ф", "F"],
      ["Х", "H"],
      ["Ц", "C"],
      ["Ч", "Ch"],
      ["Ш", "Sh"],
      ["Щ", "Sh"],
      ["Ъ", ""],
      ["Ы", "Y"],
      ["Ь", ""],
      ["Э", "E"],
      ["Ю", "Yu"],
      ["Я", "Ya"]
    ]
  },
  {
    "UKRAINIAN_MAP:uk": [
      ["Є", "Ye"],
      ["І", "I"],
      ["Ї", "Yi"],
      ["Ґ", "G"],
      ["є", "ye"],
      ["і", "i"],
      ["ї", "yi"],
      ["ґ", "g"],
      ["Г", "H"],
      ["г", "h"],
      ["И", "Y"],
      ["и", "y"],
      ["Й", "Y"],
      ["й", "i"],
      ["Х", "Kh"],
      ["х", "kh"],
      ["Ц", "Ts"],
      ["ц", "ts"],
      ["Щ", "Shch"],
      ["щ", "shch"]
    ]
  },
```

**Comment by drimacus182** [comment]
@lb- looks good to me, thanks!

**Comment by lb-** [comment]
Okie dokie. I got there in the end. Thanks for the confirmation on those unit tests @drimacus182 


I now have a PR up that implements the alternative approach proposed above 

https://github.com/wagtail/wagtail/pull/12598

It's pretty chunky, I'll be filling out the description soon after another self code review. 

It introduces two new super powers for our Django `SlugInput` widget.

1. Locale aware transliteration for slugs, when unicode slugs are disabled.
2. Formatters, a regex/replace mechanism for covering a range of use cases such as static prefixes/suffixes, bespoke character replacement, stop words removal and it all runs before the urlify logic runs. 

It builds on another PR (https://github.com/wagtail/wagtail/pull/12554) that makes the base JS `CloneController` better unit tested, fixes a few small edge cases inconsistencies and adds read only event dispatching. In addition, it should be easier to maintain & easier to override with JS class extension if the `SlugInput` super powers aren't enough. 


**нехай веселощі починаються**
